File: JsonSchemaAndStructs/KiteConnect-Python-DataClass/ticker_full.py
# This code parses date/times, so please
#
#     pip install python-dateutil
#
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = ticker_full_from_dict(json.loads(json_string))

from dataclasses import dataclass
from typing import Optional, Any, List, TypeVar, Callable, Type, cast
from datetime import datetime
import dateutil.parser


T = TypeVar("T")


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def from_none(x: Any) -> Any:
    assert x is None
    return x


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


def from_float(x: Any) -> float:
    assert isinstance(x, (float, int)) and not isinstance(x, bool)
    return float(x)


def to_float(x: Any) -> float:
    assert isinstance(x, float)
    return x


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


def from_datetime(x: Any) -> datetime:
    return dateutil.parser.parse(x)


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def from_bool(x: Any) -> bool:
    assert isinstance(x, bool)
    return x


@dataclass(slots=True)
class Buy:
    orders: Optional[int] = None
    price: Optional[float] = None
    quantity: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Buy':
        assert isinstance(obj, dict)
        orders = from_union([from_int, from_none], obj.get("orders"))
        price = from_union([from_float, from_none], obj.get("price"))
        quantity = from_union([from_int, from_none], obj.get("quantity"))
        return Buy(orders, price, quantity)

    def to_dict(self) -> dict:
        result: dict = {}
        result["orders"] = from_union([from_int, from_none], self.orders)
        result["price"] = from_union([to_float, from_none], self.price)
        result["quantity"] = from_union([from_int, from_none], self.quantity)
        return result


@dataclass(slots=True)
class Depth:
    buy: Optional[List[Buy]] = None
    sell: Optional[List[Buy]] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Depth':
        assert isinstance(obj, dict)
        buy = from_union([lambda x: from_list(Buy.from_dict, x), from_none], obj.get("buy"))
        sell = from_union([lambda x: from_list(Buy.from_dict, x), from_none], obj.get("sell"))
        return Depth(buy, sell)

    def to_dict(self) -> dict:
        result: dict = {}
        result["buy"] = from_union([lambda x: from_list(lambda x: to_class(Buy, x), x), from_none], self.buy)
        result["sell"] = from_union([lambda x: from_list(lambda x: to_class(Buy, x), x), from_none], self.sell)
        return result


@dataclass(slots=True)
class Ohlc:
    close: Optional[float] = None
    high: Optional[int] = None
    low: Optional[float] = None
    open: Optional[float] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Ohlc':
        assert isinstance(obj, dict)
        close = from_union([from_float, from_none], obj.get("close"))
        high = from_union([from_int, from_none], obj.get("high"))
        low = from_union([from_float, from_none], obj.get("low"))
        open = from_union([from_float, from_none], obj.get("open"))
        return Ohlc(close, high, low, open)

    def to_dict(self) -> dict:
        result: dict = {}
        result["close"] = from_union([to_float, from_none], self.close)
        result["high"] = from_union([from_int, from_none], self.high)
        result["low"] = from_union([to_float, from_none], self.low)
        result["open"] = from_union([to_float, from_none], self.open)
        return result


@dataclass(slots=True)
class TickerFull:
    average_price: Optional[float] = None
    buy_quantity: Optional[int] = None
    change: Optional[float] = None
    depth: Optional[Depth] = None
    instrument_token: Optional[int] = None
    last_price: Optional[float] = None
    last_quantity: Optional[int] = None
    last_trade_time: Optional[datetime] = None
    mode: Optional[str] = None
    ohlc: Optional[Ohlc] = None
    oi: Optional[int] = None
    oi_day_high: Optional[int] = None
    oi_day_low: Optional[int] = None
    sell_quantity: Optional[int] = None
    timestamp: Optional[datetime] = None
    tradable: Optional[bool] = None
    volume: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'TickerFull':
        assert isinstance(obj, dict)
        average_price = from_union([from_float, from_none], obj.get("average_price"))
        buy_quantity = from_union([from_int, from_none], obj.get("buy_quantity"))
        change = from_union([from_float, from_none], obj.get("change"))
        depth = from_union([Depth.from_dict, from_none], obj.get("depth"))
        instrument_token = from_union([from_int, from_none], obj.get("instrument_token"))
        last_price = from_union([from_float, from_none], obj.get("last_price"))
        last_quantity = from_union([from_int, from_none], obj.get("last_quantity"))
        last_trade_time = from_union([from_datetime, from_none], obj.get("last_trade_time"))
        mode = from_union([from_str, from_none], obj.get("mode"))
        ohlc = from_union([Ohlc.from_dict, from_none], obj.get("ohlc"))
        oi = from_union([from_int, from_none], obj.get("oi"))
        oi_day_high = from_union([from_int, from_none], obj.get("oi_day_high"))
        oi_day_low = from_union([from_int, from_none], obj.get("oi_day_low"))
        sell_quantity = from_union([from_int, from_none], obj.get("sell_quantity"))
        timestamp = from_union([from_datetime, from_none], obj.get("timestamp"))
        tradable = from_union([from_bool, from_none], obj.get("tradable"))
        volume = from_union([from_int, from_none], obj.get("volume"))
        return TickerFull(average_price, buy_quantity, change, depth, instrument_token, last_price, last_quantity, last_trade_time, mode, ohlc, oi, oi_day_high, oi_day_low, sell_quantity, timestamp, tradable, volume)

    def to_dict(self) -> dict:
        result: dict = {}
        result["average_price"] = from_union([to_float, from_none], self.average_price)
        result["buy_quantity"] = from_union([from_int, from_none], self.buy_quantity)
        result["change"] = from_union([to_float, from_none], self.change)
        result["depth"] = from_union([lambda x: to_class(Depth, x), from_none], self.depth)
        result["instrument_token"] = from_union([from_int, from_none], self.instrument_token)
        result["last_price"] = from_union([to_float, from_none], self.last_price)
        result["last_quantity"] = from_union([from_int, from_none], self.last_quantity)
        result["last_trade_time"] = from_union([lambda x: x.isoformat(), from_none], self.last_trade_time)
        result["mode"] = from_union([from_str, from_none], self.mode)
        result["ohlc"] = from_union([lambda x: to_class(Ohlc, x), from_none], self.ohlc)
        result["oi"] = from_union([from_int, from_none], self.oi)
        result["oi_day_high"] = from_union([from_int, from_none], self.oi_day_high)
        result["oi_day_low"] = from_union([from_int, from_none], self.oi_day_low)
        result["sell_quantity"] = from_union([from_int, from_none], self.sell_quantity)
        result["timestamp"] = from_union([lambda x: x.isoformat(), from_none], self.timestamp)
        result["tradable"] = from_union([from_bool, from_none], self.tradable)
        result["volume"] = from_union([from_int, from_none], self.volume)
        return result


def ticker_full_from_dict(s: Any) -> TickerFull:
    return TickerFull.from_dict(s)


def ticker_full_to_dict(x: TickerFull) -> Any:
    return to_class(TickerFull, x)

File: JsonSchemaAndStructs/KiteConnect-JavaScript-Structs/ticker_full.js
// To parse this data:
//
//   const Convert = require("./file");
//
//   const tickerFull = Convert.toTickerFull(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
function toTickerFull(json) {
    return cast(JSON.parse(json), r("TickerFull"));
}

function tickerFullToJson(value) {
    return JSON.stringify(uncast(value, r("TickerFull")), null, 2);
}

function invalidValue(typ, val, key = '') {
    if (key) {
        throw Error(`Invalid value for key "${key}". Expected type ${JSON.stringify(typ)} but got ${JSON.stringify(val)}`);
    }
    throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`, );
}

function jsonToJSProps(typ) {
    if (typ.jsonToJS === undefined) {
        const map = {};
        typ.props.forEach((p) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ) {
    if (typ.jsToJSON === undefined) {
        const map = {};
        typ.props.forEach((p) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val, typ, getProps, key = '') {
    function transformPrimitive(typ, val) {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val, key);
    }

    function transformUnion(typs, val) {
        // val must validate against one typ in typs
        const l = typs.length;
        for (let i = 0; i < l; i++) {
            const typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val);
    }

    function transformEnum(cases, val) {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases, val);
    }

    function transformArray(typ, val) {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue("array", val);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformDate(val) {
        if (val === null) {
            return null;
        }
        const d = new Date(val);
        if (isNaN(d.valueOf())) {
            return invalidValue("Date", val);
        }
        return d;
    }

    function transformObject(props, additional, val) {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue("object", val);
        }
        const result = {};
        Object.getOwnPropertyNames(props)
            .forEach(key => {
                const prop = props[key];
                const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
                result[prop.key] = transform(v, prop.typ, getProps, prop.key);
            });
        Object.getOwnPropertyNames(val)
            .forEach(key => {
                if (!Object.prototype.hasOwnProperty.call(props, key)) {
                    result[key] = transform(val[key], additional, getProps, key);
                }
            });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val);
    }
    if (typ === false) return invalidValue(typ, val);
    while (typeof typ === "object" && typ.ref !== undefined) {
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val) :
            typ.hasOwnProperty("arrayItems") ? transformArray(typ.arrayItems, val) :
            typ.hasOwnProperty("props") ? transformObject(getProps(typ), typ.additional, val) :
            invalidValue(typ, val);
    }
    // Numbers can be parsed by Date but shouldn't be.
    if (typ === Date && typeof val !== "number") return transformDate(val);
    return transformPrimitive(typ, val);
}

function cast(val, typ) {
    return transform(val, typ, jsonToJSProps);
}

function uncast(val, typ) {
    return transform(val, typ, jsToJSONProps);
}

function a(typ) {
    return { arrayItems: typ };
}

function u(...typs) {
    return { unionMembers: typs };
}

function o(props, additional) {
    return { props, additional };
}

function m(additional) {
    return { props: [], additional };
}

function r(name) {
    return { ref: name };
}

const typeMap = {
    "TickerFull": o([
        { json: "average_price", js: "average_price", typ: u(undefined, 3.14) },
        { json: "buy_quantity", js: "buy_quantity", typ: u(undefined, 0) },
        { json: "change", js: "change", typ: u(undefined, 3.14) },
        { json: "depth", js: "depth", typ: u(undefined, r("Depth")) },
        { json: "instrument_token", js: "instrument_token", typ: u(undefined, 0) },
        { json: "last_price", js: "last_price", typ: u(undefined, 3.14) },
        { json: "last_quantity", js: "last_quantity", typ: u(undefined, 0) },
        { json: "last_trade_time", js: "last_trade_time", typ: u(undefined, Date) },
        { json: "mode", js: "mode", typ: u(undefined, "") },
        { json: "ohlc", js: "ohlc", typ: u(undefined, r("Ohlc")) },
        { json: "oi", js: "oi", typ: u(undefined, 0) },
        { json: "oi_day_high", js: "oi_day_high", typ: u(undefined, 0) },
        { json: "oi_day_low", js: "oi_day_low", typ: u(undefined, 0) },
        { json: "sell_quantity", js: "sell_quantity", typ: u(undefined, 0) },
        { json: "timestamp", js: "timestamp", typ: u(undefined, Date) },
        { json: "tradable", js: "tradable", typ: u(undefined, true) },
        { json: "volume", js: "volume", typ: u(undefined, 0) },
    ], false),
    "Depth": o([
        { json: "buy", js: "buy", typ: u(undefined, a(r("Buy"))) },
        { json: "sell", js: "sell", typ: u(undefined, a(r("Buy"))) },
    ], false),
    "Buy": o([
        { json: "orders", js: "orders", typ: u(undefined, 0) },
        { json: "price", js: "price", typ: u(undefined, 3.14) },
        { json: "quantity", js: "quantity", typ: u(undefined, 0) },
    ], false),
    "Ohlc": o([
        { json: "close", js: "close", typ: u(undefined, 3.14) },
        { json: "high", js: "high", typ: u(undefined, 0) },
        { json: "low", js: "low", typ: u(undefined, 3.14) },
        { json: "open", js: "open", typ: u(undefined, 3.14) },
    ], false),
};

module.exports = {
    "tickerFullToJson": tickerFullToJson,
    "toTickerFull": toTickerFull,
};